[compiled autograd] support Tensor Subclasses in AOTBackward

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #144115
* #143417
* #143405
* #143387
* #143304
* #143296

Compiled autograd's initial trace traces through the AOTBackward
epilogue. The Tensor Subclass code is not traceable. This PR changes it
so that when we see Tensor Subclass constructors, we proxy nodes for
their construction into the graph.

Test Plan:
- New basic test with TwoTensor
- Existing tests

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov @xmfan

zou3519 committed:
[compiled autograd] support Tensor Subclasses in AOTBackward

Compiled autograd's initial trace traces through the AOTBackward
epilogue. The Tensor Subclass code is not traceable. This PR changes it
so that when we see Tensor Subclass constructors, we proxy nodes for
their construction into the graph.

We also change a DTensor constructor call to be DTensor.from_local. This
is because Dynamo cannot handle raw DTensor constructor calls but is
able to handle DTensor.from_local.

Test Plan:
- More compiled autograd x Tensor subclass tests pass now. The other
  failures need investigation but a lot of them are because Dynamo
  isn't able to trace DTensor internals that are now exposed (due to
  autograd.Function).
- Existing tests

[ghstack-poisoned]

zou3519 committed:
Update on "[compiled autograd] support Tensor Subclasses in AOTBackward"


Compiled autograd's initial trace traces through the AOTBackward
epilogue. The Tensor Subclass code is not traceable. This PR changes it
so that when we see Tensor Subclass constructors, we proxy nodes for
their construction into the graph.

We also change a DTensor constructor call to be DTensor.from_local. This
is because Dynamo cannot handle raw DTensor constructor calls but is
able to handle DTensor.from_local.

Test Plan:
- More compiled autograd x Tensor subclass tests pass now. The other
  failures need investigation but a lot of them are because Dynamo
  isn't able to trace DTensor internals that are now exposed (due to
  autograd.Function).
- Existing tests

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov xmfan

[ghstack-poisoned]

zou3519 committed:
Update on "[compiled autograd] support Tensor Subclasses in AOTBackward"


Compiled autograd's initial trace traces through the AOTBackward
epilogue. The Tensor Subclass code is not traceable. This PR changes it
so that when we see Tensor Subclass constructors, we proxy nodes for
their construction into the graph.

NB: I tried to run this on the DTensor x compiled autograd tests, but it
didn't improve anything. The main problem there was that the DTensor
internal code now needs to be Dynamo-able and it is not (there are
things like DTensor constructor calls and attribute accesses that are
not yet Dynamo-able).

Test Plan:
- New basic test with TwoTensor
- Existing tests

[ghstack-poisoned]

zou3519 committed:
Update on "[compiled autograd] support Tensor Subclasses in AOTBackward"


Compiled autograd's initial trace traces through the AOTBackward
epilogue. The Tensor Subclass code is not traceable. This PR changes it
so that when we see Tensor Subclass constructors, we proxy nodes for
their construction into the graph.

Test Plan:
- New basic test with TwoTensor
- Existing tests

[ghstack-poisoned]

zou3519 committed:
Update on "[compiled autograd] support Tensor Subclasses in AOTBackward"


Compiled autograd's initial trace traces through the AOTBackward
epilogue. The Tensor Subclass code is not traceable. This PR changes it
so that when we see Tensor Subclass constructors, we proxy nodes for
their construction into the graph.

Test Plan:
- New basic test with TwoTensor
- Existing tests

[ghstack-poisoned]

zou3519 committed:
Update on "[compiled autograd] support Tensor Subclasses in AOTBackward"


Compiled autograd's initial trace traces through the AOTBackward
epilogue. The Tensor Subclass code is not traceable. This PR changes it
so that when we see Tensor Subclass constructors, we proxy nodes for
their construction into the graph.

Test Plan:
- New basic test with TwoTensor
- Existing tests

[ghstack-poisoned]

zou3519 committed:
Update on "[compiled autograd] support Tensor Subclasses in AOTBackward"


Compiled autograd's initial trace traces through the AOTBackward
epilogue. The Tensor Subclass code is not traceable. This PR changes it
so that when we see Tensor Subclass constructors, we proxy nodes for
their construction into the graph.

Test Plan:
- New basic test with TwoTensor
- Existing tests

[ghstack-poisoned]

zou3519 commented on test/inductor/test_compiled_autograd.py:
```diff
@@ -3133,6 +3134,122 @@ def fn():
+        getitem_3: "f32[s0, s0]" = inputs[3]
+        getitem_4: "f32[s0, s0]" = inputs[4];  inputs = None
+
+        validate_outputs = torch__dynamo_compiled_autograd_ops_validate_outputs([getitem], [((None, None, device(type='cpu'), 6, 0, None), [], True)]);  getitem = None
+        getitem_5: "f32[]" = validate_outputs[0];  validate_outputs = None
+
+        sum_backward0 = torch__dynamo_compiled_autograd_ops_SumBackward0([getitem_5], [True], [4, 4]);  getitem_5 = None
+        getitem_6: "f32[0, 123456789]" = sum_backward0[0];  sum_backward0 = None
+        validate_outputs_1 = torch__dynamo_compiled_autograd_ops_validate_outputs([getitem_6], [((None, None, device(type='cpu'), 6, 0, None), [4, 4], True)]);  getitem_6 = None
+        getitem_7: "f32[0, 123456789]" = validate_outputs_1[0];  validate_outputs_1 = None
```
TODO(rzou): I don't think I got rid of the metadata correctly in the previous PR

xmfan commented on test/inductor/test_compiled_autograd.py:
```diff
@@ -3133,6 +3134,122 @@ def fn():
+        call_backward = torch__dynamo_external_utils_call_backward(getitem_14, (), getitem_13);  getitem_14 = getitem_13 = None
+        getitem_17: "f32[4, 4]" = call_backward[0]
+        getitem_18: "f32[4, 4]" = call_backward[1];  call_backward = None
+        validate_outputs_3 = torch__dynamo_compiled_autograd_ops_validate_outputs([getitem_17, getitem_18], [((None, None, device(type='cpu'), 6, 0, None), [4, 4], False), ((None, None, device(type='cpu'), 6, 0, None), [4, 4], False)]);  getitem_17 = getitem_18 = None
```
isn't this validate_output supposed to be removed

zou3519 replied:
Was removed

zou3519 committed:
Update on "[compiled autograd] support Tensor Subclasses in AOTBackward"


Compiled autograd's initial trace traces through the AOTBackward
epilogue. The Tensor Subclass code is not traceable. This PR changes it
so that when we see Tensor Subclass constructors, we proxy nodes for
their construction into the graph.

Test Plan:
- New basic test with TwoTensor
- Existing tests

[ghstack-poisoned]

zou3519 committed:
Update on "[compiled autograd] support Tensor Subclasses in AOTBackward"


Compiled autograd's initial trace traces through the AOTBackward
epilogue. The Tensor Subclass code is not traceable. This PR changes it
so that when we see Tensor Subclass constructors, we proxy nodes for
their construction into the graph.

Test Plan:
- New basic test with TwoTensor
- Existing tests

[ghstack-poisoned]

zou3519 committed:
Update on "[compiled autograd] support Tensor Subclasses in AOTBackward"


Compiled autograd's initial trace traces through the AOTBackward
epilogue. The Tensor Subclass code is not traceable. This PR changes it
so that when we see Tensor Subclass constructors, we proxy nodes for
their construction into the graph.

Test Plan:
- New basic test with TwoTensor
- Existing tests

[ghstack-poisoned]

zou3519 committed:
Update on "[compiled autograd] support Tensor Subclasses in AOTBackward"


Compiled autograd's initial trace traces through the AOTBackward
epilogue. The Tensor Subclass code is not traceable. This PR changes it
so that when we see Tensor Subclass constructors, we proxy nodes for
their construction into the graph.

Test Plan:
- New basic test with TwoTensor
- Existing tests

[ghstack-poisoned]

zou3519 committed:
Update on "[compiled autograd] support Tensor Subclasses in AOTBackward"


Compiled autograd's initial trace traces through the AOTBackward
epilogue. The Tensor Subclass code is not traceable. This PR changes it
so that when we see Tensor Subclass constructors, we proxy nodes for
their construction into the graph.

Test Plan:
- New basic test with TwoTensor
- Existing tests

[ghstack-poisoned]

zou3519 committed:
Update on "[compiled autograd] support Tensor Subclasses in AOTBackward"


Compiled autograd's initial trace traces through the AOTBackward
epilogue. The Tensor Subclass code is not traceable. This PR changes it
so that when we see Tensor Subclass constructors, we proxy nodes for
their construction into the graph.

Test Plan:
- New basic test with TwoTensor
- Existing tests

[ghstack-poisoned]

zou3519 commented:
@pytorchbot merge

zou3519 committed:
Update on "[compiled autograd] support Tensor Subclasses in AOTBackward"


Compiled autograd's initial trace traces through the AOTBackward
epilogue. The Tensor Subclass code is not traceable. This PR changes it
so that when we see Tensor Subclass constructors, we proxy nodes for
their construction into the graph.

Test Plan:
- New basic test with TwoTensor
- Existing tests

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov xmfan

[ghstack-poisoned]

zou3519 committed:
Update on "[compiled autograd] support Tensor Subclasses in AOTBackward"


Compiled autograd's initial trace traces through the AOTBackward
epilogue. The Tensor Subclass code is not traceable. This PR changes it
so that when we see Tensor Subclass constructors, we proxy nodes for
their construction into the graph.

Test Plan:
- New basic test with TwoTensor
- Existing tests

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov xmfan

[ghstack-poisoned]
